### PR TITLE
chore: add bazel caching to integration tests

### PR DESCRIPTION
**Pull request by software-dov**
The integration test CI takes ~8 minutes to run, most of which is
installing bazel and setting up tool dependencies.

It is a massive time optimization to cache this work and prevent its
recreation on subsequent builds.

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -312,32 +312,35 @@ jobs:
...
-      - name: Install system dependencies.
+          path: ~/.cache/bazel
+          key: ${{ runner.os }}-bazel-20210105-${{ secrets.CACHE_VERSION }}
+      - name: Cache not found
+        if: steps.cache-bazel.outputs.cache-hit != 'true'
         run: |
-          sudo apt-get update
-          sudo apt-get install -y curl pandoc unzip gcc
-      - name: Install Bazel
+          echo "No cache found."
```
Comment (tseaver) [inline .github/workflows/tests.yaml:331]: Shouldn't a cache miss imply that we actually do something to populate the missed files?

Comment (software-dov) [reply .github/workflows/tests.yaml:331]: This is done implicitly by bazel during startup. It checks for the existence of `~/.cache/bazel` and if not found runs all the tedious startup of compiling `gRPC`, `protoc`, etc. which are then put into the cache.
The github action saves this directory persistently. At the start of a CI run, it checks for that directory in persistent storage and, if found, populates it in the container.

Comment (software-dov) [reply .github/workflows/tests.yaml:331]: Anecdotally, this cuts the integration test down to about 1 minute 30 seconds.

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -312,32 +312,35 @@ jobs:
...
-      - name: Set up Python 3.8
-        uses: actions/setup-python@v2
+      - name: Cache Bazel files
+        id: cache-bazel
+        uses: actions/cache@v2
         with:
-          python-version: 3.8
-      - name: Install system dependencies.
+          path: ~/.cache/bazel
+          key: ${{ runner.os }}-bazel-20210105-${{ secrets.CACHE_VERSION }}
```
Comment (busunkim96) [inline .github/workflows/tests.yaml:327]: Does the '20210105' bit ever need to be updated? If so, please explain in a comment.

**Review by busunkim96 — APPROVED** [review]
LGTM (with the caveat that I don't know much about Bazel 👀 )

**Pull request merged**